tools/python/gen_opkernel_doc.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python
+
+from collections import defaultdict
+import io
+import os
+import sys
+import argparse
+
+
+import onnxruntime as rt
+import onnxruntime.capi.onnxruntime_pybind11_state as rtpy 
+from onnxruntime.capi.onnxruntime_pybind11_state import opkernel 
+from onnxruntime.capi.onnxruntime_pybind11_state.opkernel import KernelDef 
+
+
+def format_version_range(v):
+    if (v[1] >= 2147483647):
+        return str(v[0])+'+'
+    else:
+        return '['+str(v[0])+', '+str(v[1])+']'    
+
+def format_type_constraints(tc):
+    counter = 0
+    tcstr = ''
+    firsttcitem = True
+    for tcitem in tc:
+        counter += 1
+        if firsttcitem:
+            firsttcitem = False
+        else:
+            tcstr += ', '
+        tcstr += tcitem
+    return tcstr
+    
+def main(args):  # type: (Type[Args]) -> None
+    
+    with io.open(args.output, 'w', newline='', encoding="utf-8") as fout:
+        fout.write('## Supported Operators Data Types\n')
+        fout.write(
+            "*This file is automatically generated from the\n"
+            "            [def files](/onnxruntime/core/providers/cpu/cpu_execution_provider.cc) via [this script](/tools/python/gen_opkernel_doc.py).\n"
+            "            Do not modify directly and instead edit operator definitions.*\n")
+
+        index = defaultdict(lambda: defaultdict(lambda: defaultdict(list))) 
+        for op in rtpy.get_all_opkernel_def():
+            domain = op.domain
+            if (domain == ''):
+                domain = 'ai.onnx.ml'
+            index[op.provider][domain][op.op_name].append(op)
+
+               
+        fout.write('\n')
+        for provider, domainmap in sorted(index.items()):
+            fout.write('\n\n## Operators implemented by '+provider+'\n\n')
+            fout.write('| Op Name | OpSet Versions | Parameter | Types Supported |\n')
```

Comment:
The parameter column shows "T" for many ops. T is the placeholder for the type. I think it should show the actual parameter name like X/Y/Z.

Reply:
The type name 'T' comes from the TypeConstraints map of the KernelDef object. Typically it is 'T' when there is a single input parameter. Not sure how the KernelDef is populated. If we want to change the name, we need to populate there.

Reply:
Yeah, I see that. Looks like when we register the kernels we use type ids like T, T1, etc instead of the names. This makes it difficult to say which types are supported for which inputs/outputs. Other than fixing the registration of all the kernels which will be quite painful and can break in the future, I'm wondering if there's another way.

Reply:
retrieved the operator input/output params from the opschema.